[BUG] false positive for Eventually/Consistently

**Describe the bug**

The following finding is a false positive:

```
test/e2e_node/dra_test.go:129:4: ginkgo-linter: use a function call in Consistently. This actually checks nothing, because Consistently receives the function returned value, instead of function itself, and this value is never changed; consider using `gomega.Consistently(ctx, e2epod.Get).WithArguments(f.ClientSet, pod).WithTimeout(podInPendingStateTimeout).Should(e2epod.BeInPhase(v1.PodPending),
	"Pod should be in Pending state as resource preparation time outed")` instead (ginkgolinter)
			gomega.Consistently(ctx, e2epod.Get(f.ClientSet, pod)).WithTimeout(podInPendingStateTimeout).Should(e2epod.BeInPhase(v1.PodPending),
			^
```

It's a false positive because `e2epod.Get` is a function call which returns a function. That function gets called by `Consistently` repeatedly.

**Expected behavior**

I suspect this occurs because ginkgolinter checks whether the parameter is a function call expression. What it should do instead is check what the *type* of the expression is. Functions and channels are okay, everything else isn't.

**Environment:**
 - OS: Linux
 - Version: golangci-lint 1.53.3

**Additional context**

This can be worked around:

```patch
diff --git a/test/e2e_node/dra_test.go b/test/e2e_node/dra_test.go
index fb3da210bf2..6ea97e91a1d 100644
--- a/test/e2e_node/dra_test.go
+++ b/test/e2e_node/dra_test.go
@@ -126,7 +126,8 @@ var _ = ginkgo.Describe("[sig-node] DRA [Feature:DynamicResourceAllocation][Node
 			// TODO: Check condition or event when implemented
 			// see https://github.com/kubernetes/kubernetes/issues/118468 for details
 			ginkgo.By("check that pod is consistently in Pending state")
-			gomega.Consistently(ctx, e2epod.Get(f.ClientSet, pod)).WithTimeout(podInPendingStateTimeout).Should(e2epod.BeInPhase(v1.PodPending),
+			get := e2epod.Get(f.ClientSet, pod)
+			gomega.Consistently(ctx, get).WithTimeout(podInPendingStateTimeout).Should(e2epod.BeInPhase(v1.PodPending),
 				"Pod should be in Pending state as resource preparation time outed")
 		})
 	})
```

--- a/ginkgo_linter.go
+++ b/ginkgo_linter.go
@@ -333,10 +333,7 @@ func checkAsyncAssertion(pass *analysis.Pass, config types.Config, expr *ast.Cal
 		if len(actualExpr.Args) > funcIndex {
 			if fun, funcCall := actualExpr.Args[funcIndex].(*ast.CallExpr); funcCall {
 				t = pass.TypesInfo.TypeOf(fun)
-				switch t.(type) {
-				// allow functions that return function or channel.
-				case *gotypes.Signature, *gotypes.Chan, *gotypes.Pointer:
-				default:
+				if !isValidAsyncValueType(t) {
 					actualExpr = handler.GetActualExpr(expr.Fun.(*ast.SelectorExpr))
 
 					if len(fun.Args) > 0 {
@@ -369,6 +366,18 @@ func checkAsyncAssertion(pass *analysis.Pass, config types.Config, expr *ast.Cal
 
 	handleAssertionOnly(pass, config, expr, handler, actualExpr, oldExpr, true)
 	return true
+}
+
+func isValidAsyncValueType(t gotypes.Type) bool {
+	switch t.(type) {
+	// allow functions that return function or channel.
+	case *gotypes.Signature, *gotypes.Chan, *gotypes.Pointer:
+		return true
+	case *gotypes.Named:
+		return isValidAsyncValueType(t.Underlying())
+	}
+
+	return false
 }
 
 func startCheckComparison(exp *ast.CallExpr, handler gomegahandler.Handler) (*ast.CallExpr, bool) {

--- a/testdata/src/a/eventually/eventuallyFunc.go
+++ b/testdata/src/a/eventually/eventuallyFunc.go
@@ -9,6 +9,14 @@ import (
 	. "github.com/onsi/gomega"
 )
 
+type GetFunc[T any] func(ctx context.Context) (T, error)
+
+func retFuncType() GetFunc[int] {
+	return func(_ context.Context) (int, error) {
+		return 1, nil
+	}
+}
+
 func slowInt() int {
 	time.Sleep(time.Second)
 	return 42
@@ -46,6 +54,10 @@ func retChan() chan int {
 
 var _ = Describe("Eventually with function", func() {
 	Context("Eventually", func() {
+		It("should not trigger warning", func() {
+			Eventually(retFuncType()).Should(Equal(1))
+		})
+
 		Eventually(retFunc(5)).Should(Equal(5)) // valid. retFunc returns a function
 		ch := make(chan int)
 		Eventually(ch).Should(BeClosed())                                                                                        // valid

